File: PYTHON-DSA/singly_linked_list/pop.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None


class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def push(self, val):
        new_node = Node(val)
        if not self.head:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node
        self.length += 1
        return self

    def pop(self):
        if not self.head:
            return None

        current = self.head
        new_tail = current

        while current.next:
            new_tail = current
            current = current.next

        self.tail = new_tail
        self.tail.next = None
        self.length -= 1

        if (self.length == 0):
            this.head = None
            this.tail = None

        return current

    def print_list(self):
        current = self.head
        while current:
            print(f"{current.val} -> ", end="")
            current = current.next
        print("NULL")


list = SinglyLinkedList()
list.push("Hello")
list.push("Goodbye")
list.push("!")
list.print_list()

print(":::::::::::::::::")

list.pop()
list.print_list()

File: GO-DSA/singly_linked_list/pop.go
package main

import (
	"fmt"
)

type Node struct {
	val  string
	next *Node
}

type SingleLinkedList struct {
	head   *Node
	tail   *Node
	length int
}

func (s *SingleLinkedList) Push(val string) {
	newNode := &Node{val: val}
	if s.head == nil {
		s.head = newNode
		s.tail = newNode
	} else {
		s.tail.next = newNode
		s.tail = newNode
	}

	s.length++

}

func (s *SingleLinkedList) Pop() *Node {
	if s.head == nil {
		return nil
	}

	current := s.head
	newTail := current

	for current.next != nil {
		newTail = current
		current = current.next
	}

	s.tail = newTail
	s.tail.next = nil
	s.length--

	if s.length == 0 {
		s.head = nil
		s.tail = nil
	}

	return current

}

func (s *SingleLinkedList) PrintList() {
	current := s.head
	for current != nil {
		fmt.Printf("%s -> ", current.val)
		current = current.next
	}

	fmt.Println("NULL")
}

func main() {
	list := &SingleLinkedList{}
	list.Push("Hello")
	list.Push("Goodbye")
	list.Push("!")
	list.PrintList()

	list.Pop()
	list.PrintList()
}
